File: backend/src/sor/models/stage.py
from __future__ import annotations

from datetime import datetime, timezone
from enum import StrEnum
from uuid import uuid4

from pydantic import BaseModel, Field

from .agent import AgentConfig, AgentOutput
from .conflict import ConflictReport


def _new_id() -> str:
    return uuid4().hex[:12]


def _now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()


class StageStatus(StrEnum):
    PENDING = "pending"
    RUNNING = "running"
    COMPLETE = "complete"
    APPROVED = "approved"
    SKIPPED = "skipped"


STAGE_NAMES: dict[int, str] = {
    1: "Problem Framing",
    2: "Evidence Gathering",
    3: "Analysis & Interpretation",
    4: "Insight Synthesis",
    5: "Communication",
    6: "Prototype & Intervention Design",
}

STAGE_DESCRIPTIONS: dict[int, str] = {
    1: "Define the research question, scope, and stakeholder landscape.",
    2: "Identify and gather evidence from multiple sources.",
    3: "Analyze evidence through multiple lenses, surface patterns and conflicts.",
    4: "Synthesize findings into actionable insights with confidence levels.",
    5: "Generate tailored deliverables for different audiences.",
    6: "Design interventions and prototypes grounded in research findings.",
}


class StageDefinition(BaseModel):
    number: int
    name: str
    description: str
    agents: list[AgentConfig] = Field(default_factory=list)


class StageResult(BaseModel):
    id: str = Field(default_factory=_new_id)
    project_id: str
    stage_number: int
    status: StageStatus = StageStatus.PENDING
    agent_outputs: list[AgentOutput] = Field(default_factory=list)
    conflict_report: ConflictReport | None = None
    human_override: str | None = None
    human_notes: str = ""
    approved_at: str | None = None
    created_at: str = Field(default_factory=_now_iso)

File: ios/SocietyOfResearchers/Models/Stage.swift
import Foundation

enum StageStatus: String, Codable, Sendable {
    case pending
    case running
    case complete
    case approved
    case skipped
}

struct StageResult: Codable, Identifiable, Sendable {
    let id: String
    let projectId: String
    let stageNumber: Int
    var status: StageStatus
    var agentOutputs: [AgentOutput]
    var conflictReport: ConflictReport?
    var humanOverride: String?
    var humanNotes: String
    var approvedAt: String?
    var createdAt: String

    enum CodingKeys: String, CodingKey {
        case id, status
        case projectId = "project_id"
        case stageNumber = "stage_number"
        case agentOutputs = "agent_outputs"
        case conflictReport = "conflict_report"
        case humanOverride = "human_override"
        case humanNotes = "human_notes"
        case approvedAt = "approved_at"
        case createdAt = "created_at"
    }
}

struct StageInfo: Sendable {
    let number: Int
    let name: String
    let description: String

    static let all: [StageInfo] = [
        StageInfo(number: 1, name: "Problem Framing", description: "Define the research question, scope, and stakeholder landscape."),
        StageInfo(number: 2, name: "Evidence Gathering", description: "Identify and gather evidence from multiple sources."),
        StageInfo(number: 3, name: "Analysis & Interpretation", description: "Analyze evidence through multiple lenses, surface patterns and conflicts."),
        StageInfo(number: 4, name: "Insight Synthesis", description: "Synthesize findings into actionable insights with confidence levels."),
        StageInfo(number: 5, name: "Communication", description: "Generate tailored deliverables for different audiences."),
        StageInfo(number: 6, name: "Prototype & Intervention Design", description: "Design interventions and prototypes grounded in research findings."),
    ]

    static func name(for number: Int) -> String {
        all.first { $0.number == number }?.name ?? "Stage \(number)"
    }

    static func description(for number: Int) -> String {
        all.first { $0.number == number }?.description ?? ""
    }
}

struct ApproveResponse: Decodable, Sendable {
    let ok: Bool
    var complete: Bool?
    var nextStage: Int?

    enum CodingKeys: String, CodingKey {
        case ok, complete
        case nextStage = "next_stage"
    }
}

struct OverrideRequest: Encodable, Sendable {
    let content: String
    var notes: String = ""
}
